src/WebJobs.Script/Host/ProxyMetadataManager.cs

```diff
@@ -7,33 +7,74 @@
 using System.Collections.ObjectModel;
 using System.IO;
 using System.Linq;
+using System.Reactive.Linq;
 using System.Text.RegularExpressions;
+using System.Threading;
 using Microsoft.Azure.AppService.Proxy.Client;
 using Microsoft.Azure.WebJobs.Logging;
 using Microsoft.Azure.WebJobs.Script.Description;
+using Microsoft.Azure.WebJobs.Script.Eventing;
 using Microsoft.Extensions.Logging;
 using Microsoft.Extensions.Options;
 using Newtonsoft.Json.Linq;
 
 namespace Microsoft.Azure.WebJobs.Script
 {
-    public class ProxyMetadataManager : IProxyMetadataManager
+    public class ProxyMetadataManager : IProxyMetadataManager, IDisposable
     {
         private static readonly Regex ProxyNameValidationRegex = new Regex(@"[^a-zA-Z0-9_-]", RegexOptions.Compiled | RegexOptions.IgnoreCase);
-        private readonly Lazy<ProxyMetadataInfo> _metadata;
+        private readonly ReaderWriterLockSlim _metadataLock = new ReaderWriterLockSlim();
         private readonly IOptionsMonitor<ScriptApplicationHostOptions> _applicationHostOptions;
         private readonly IEnvironment _environment;
         private readonly ILogger _logger;
+        private readonly IDisposable _fileChangeSubscription;
+        private Lazy<ProxyMetadataInfo> _metadata;
+        private bool _disposed = false;
 
-        public ProxyMetadataManager(IOptionsMonitor<ScriptApplicationHostOptions> applicationHostOptions, IEnvironment environment, ILoggerFactory loggerFactory)
+        public ProxyMetadataManager(IOptionsMonitor<ScriptApplicationHostOptions> applicationHostOptions, IEnvironment environment, IScriptEventManager eventManager, ILoggerFactory loggerFactory)
         {
             _applicationHostOptions = applicationHostOptions;
             _environment = environment;
             _logger = loggerFactory.CreateLogger(LogCategories.Startup);
             _metadata = new Lazy<ProxyMetadataInfo>(LoadFunctionMetadata);
+
+            _fileChangeSubscription = eventManager.OfType<FileEvent>()
+                       .Where(f => string.Equals(f.Source, EventSources.ScriptFiles, StringComparison.Ordinal) &&
+                       string.Equals(Path.GetFileName(f.FileChangeArguments.Name), ScriptConstants.ProxyMetadataFileName, StringComparison.OrdinalIgnoreCase))
+                       .Subscribe(e => HandleProxyFileChange());
```

Comment:
Yes, this was a concern and one of the reasons the subscription doesn't debounce the handler here as well (wanted to avoid deferring this work, principally in scenarios where we're short in resources). The order is dependent on subscription and the scheduler. Currently, this shouldn't be an issue because of the way those events are processed by the file monitor, but it is something that was bothering me. I have some thoughts on this and will follow up with general improvements. Didn't want to do that as part of this change as it would increase the scope and since we want to have this in ASAP.